kivy/properties.pyx

```diff
@@ -635,18 +635,16 @@ cdef class NumericProperty(Property):
     cdef check(self, EventDispatcher obj, value):
         if Property.check(self, obj, value):
             return True
-        if type(value) not in (int, float, long):
-            raise ValueError('%s.%s accept only int/float/long (got %r)' % (
+        if not isinstance(value, (int, float, long)):
+            raise ValueError('%s.%s accept only instances of int/float/long'
+                             '(got %r)' % (
                 obj.__class__.__name__,
                 self.name, value))
 
     cdef convert(self, EventDispatcher obj, x):
-        if x is None:
+        if x is None or isinstance(x, (int, float, long)):
             return x
-        tp = type(x)
-        if tp is int or tp is float or tp is long:
-            return x
-        if tp is tuple or tp is list:
+        if isinstance(x, (tuple, list)):
```

Comment:
I would not test for `tuple`/`list` as it can be any collection of size two, but write to test for string value first with single `raise` statement like in `ColorProperty` https://github.com/kivy/kivy/blob/master/kivy/properties.pyx#L2020 .

Reply:
> I would not test for `tuple`/`list` as it can be any collection of size two, but write to test for string value first with single `raise` statement like in `ColorProperty` https://github.com/kivy/kivy/blob/master/kivy/properties.pyx#L2020 .

I get your point, but I would not check for `string_types`, because we
already dropped the Python 2 support.

These changes are not directly related to the pull request title and
objective, so I will rename it and make some changes.

Thanks for the review, I will let you know when it is ready.

Reply:
> > The usecase was something like, there is a widget displaying a numeric value (say, win rate). The .kv handles setting the .text given the NumericProperty value. The value comes from some machine learning model.
> 
> Why can't you use just a object property? Or maybe convert it to a float before setting it. Or even a string property since you'll be converting it to a string. Either way, there's no benefit in storing it as a numeric property. I think the `NumericProperty` is a tempting distraction in this case, because it's intended for places where you need float/ints as mentioned, and it's mainly useful for internal kivy stuff to convert from e.g. `'5dp'` to a number. For normal stuff I'd highly recommend to use object property instead of string/numeric property.
> 
> Perhaps we need to say this in the docs.
> 
> > It would have been fine to raise a TypeError and tell me to cast it to float (although accepting the numpy float would be nicer, of course). The thing that took quite a bit of time was the error about an 'invalid format'.
> 
> Yes, the error should be clearer. And we should ideally convert anything convertible to a float/int to float/int and only then raise an error if it doesn't work.

I get your point about using an ObjectProperty for this use case
@matham, but it is not clear at first sight. When you are using a
numeric type, the first thing that comes to mind is to use a
NumericProperty instead of an ObjectProperty.

So, to summarize (and solve this):

- We should add these use cases in the docs to let people know when they
  should use ObjectProperty instead of NumericProperty.

- The TypeError must be more explicit. Ideally, it should suggest the
  use of ObjectProperty.

- `long` must be removed as it is deprecated.

As I mentioned before, I do not think we should cast the numpy numbers
to float/ints because it is not the idea behind using these. They might
be useful for mathematical computation without losing performance and
speed.

Reply:
I'll try to make the case for converting to float/int. As I mentioned, and you seem to kinda agree now(?) that `NumericProperty` is not a general numeric property but intended for widget sizing/pos etc. 

I kinda said this before, but it's even stronger than that. Its primary purpose seemingly is meant to take the input and convert it to something that can be used by the layout/graphics system. Hence why it converts strings to numbers using `dpi`. This is why we do all the type checking to make sure we only end up with float/int currently. Hence why I really think that if we were to broaden the input, they still will need to output either a float or int, otherwise it may end up breaking the layout/graphics system because you'd pass it a numpy number that is not understood by gl.

It is unfortunate that we named it generically `NumericProperty`, as it does make it seem like it's accepting a generic number and I'll admit to using it wherever I use numbers in the past. But that's not intended really the case I think. Actually, even the performance point is contrary to your argument. That's because `NumericProperty` does all this additional type checking, which `ObjectProperty` does not. So if you were worried about performance of setting the property, I'd definitely go with a `ObjectProperty` which doesn't do any checking (except for equality).

That said, I'm open to changing my mind about this, but would need a stronger argument of actual use cases for not converting. And would like to hear if e.g. @tshirtman thinks differently as he has much more experience with this. Either way, I agree about the docs having to be much clearer about intended usage and your other bullet points.

Reply:
Sorry for not answering earlier.
The consideration here is wide, do we want the kivy property system to be useful outside of kivy's internal management? It seems we use it for other things when we do kivy apps, when the technical constraints it imposes allow it, so there is a case for it. But do we want to relax some of these constraints to allow more use cases? It seems like a positive thing at first, but on the other hand, it will add more complexity to the codebase, as we will have to care about cases that didn't exist before, and we must check the failure modes when these new usages are wrongly applied to the main kivy usage, e.g, what happens if someone decides to pass a complex number as the width of a widget? Surely there is an error, but is it an acceptable one? Another consideration is performance, the added complexity might slow down the processing, and considering how important the performance of NumericProperty is to Kivy, that might be a cost we don't want to bear, despite the convenience of (ab)using kivy property system outside of its originaly intended use.